path.go

```diff
@@ -0,0 +1,133 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package logicalcluster
+
+import (
+	"encoding/json"
+	"path"
+	"regexp"
+	"strings"
+)
+
+const (
+
+	// ClusterHeader set to "<lcluster>" on a request is an alternative to accessing the
+	// cluster via /clusters/<lcluster>. With that the <lcluster> can be access via normal kube-like
+	// /api and /apis endpoints.
+	ClusterHeader = "X-Kubernetes-Cluster"
+
+	separator = ":"
+)
+
+// Path represents a logical cluster. A logical cluster is
+//  1. a (part of) etcd prefix to store objects in that cluster
+//  2. a (part of) a http path which serves a Kubernetes-cluster-like API with
+//     discovery, OpenAPI and the actual API groups.
+//  3. a value in metadata.clusterName in objects from cross-workspace list/watches,
+//     which is used to identify the logical cluster.
+//
+// A logical cluster is a colon separated list of words. In other words, it is
+// like a path, but with colons instead of slashes.
+type Path struct {
+	value string
+}
+
+// New returns a Path from a string.
+func New(value string) Path {
+	return Path{value}
+}
+
+// NewValidated returns a Path from a string and whether it is a valid logical cluster.
+// A valid logical cluster returns true on IsValid().
+func NewValidated(value string) (Path, bool) {
+	n := Path{value}
+	return n, n.IsValid()
+}
+
+// Empty returns true if the logical cluster value is unset.
+func (n Path) Empty() bool {
+	return n.value == ""
+}
+
+// Path returns a path segment for the logical cluster to access its API.
+func (n Path) Path() string {
+	return path.Join("/clusters", n.value)
+}
+
+// String returns the string representation of the logical cluster name.
+func (n Path) String() string {
+	return n.value
+}
+
+// Parent returns the parent logical cluster name of the given logical cluster name.
+func (n Path) Parent() (Path, bool) {
+	parent, _ := n.Split()
+	return parent, parent.value != ""
+}
+
+// Split splits logical cluster immediately following the final colon,
+// separating it into a parent logical cluster and name component.
+// If there is no colon in path, Split returns an empty logical cluster name
+// and name set to path.
+func (n Path) Split() (parent Path, name string) {
+	i := strings.LastIndex(n.value, separator)
+	if i < 0 {
+		return Path{}, n.value
+	}
+	return Path{n.value[:i]}, n.value[i+1:]
+}
+
+// Base returns the last component of the logical cluster name.
+func (n Path) Base() string {
```

Comment:
does it become `func (p Path) Name() Name` ?

Reply:
no. Normally, this does not even return a logical cluster name. Only the first element in a path is usually (not necessarily, there might be exceptions) a cluster name.

Reply:
does it mean we need a new `func (p Path) Name() Name` method that simply takes the whole path as a `Name` i.e. `Name{path.value}`

Reply:
`func (p Path) Name() (Name, bool)`